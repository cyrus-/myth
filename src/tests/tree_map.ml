let s = "
type nat =
  | O
  | S of nat

type list =
  | Nil
  | Cons of nat * list

type tree =
  | Leaf
  | Node of tree * nat * tree

let rec div2 (n:nat) : nat =
  match n with
  | O -> O
  | S (n1) -> match n1 with
    | O -> O
    | S (n2) -> S (div2 n2)
;;

let rec inc (n:nat) : nat =
  S( n )
;;


let tree_map : (nat -> nat) -> tree -> tree |>
{ div2 => ( Leaf => Leaf
          | Node (Leaf, 0, Leaf) => Node (Leaf, 0, Leaf)
          | Node (Leaf, 2, Leaf) => Node (Leaf, 1, Leaf)
          | Node (Node (Leaf, 2, Leaf), 2, Leaf) => Node (Node (Leaf, 1, Leaf), 1, Leaf) 
          | Node (Leaf, 1, Node (Leaf, 2, Leaf)) => Node (Leaf, 0, Node (Leaf, 1, Leaf)) )
| inc =>  ( Leaf => Leaf
          | Node (Leaf, 0, Leaf) => Node (Leaf, 1, Leaf) )
} = ?"
