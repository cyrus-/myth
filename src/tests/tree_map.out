(* Generated Thu Nov 13 00:35:54 2014 *)
let tree_map : (nat -> nat) -> tree -> tree =
  fun (f2:nat -> nat) ->
    let rec f3 (t1:tree) : tree =
      match t1 with
        | Leaf -> Leaf
        | Node (t2, n1, t3) -> Node (f3 t2, f2 n1, f3 t3)
    in
      f3
;;
